File: bindings/typescript/packages/aster/src/trust/consumer.ts
/**
 * Consumer admission — client-side and server-side admission handshake.
 *
 * Spec reference: Aster-trust-spec.md S5
 *
 * Client side (performAdmission):
 *   1. Open a stream on the admission ALPN
 *   2. Send a ConsumerAdmissionRequest (credential + optional IID token)
 *   3. Receive a ConsumerAdmissionResponse (services list + registry namespace)
 *
 * Server side (handleConsumerAdmissionRpc, serveConsumerAdmission):
 *   Verify credential, admit peer, return response with services + registry namespace.
 *
 * Wire format: newline-delimited JSON over a QUIC bidi-stream on
 * aster.consumer_admission ALPN. Client sends one JSON line; server
 * responds with one JSON line and closes the stream.
 */

import { bytesToHex, type ConsumerEnrollmentCredential } from './credentials.js';
import { admit } from './admission.js';
import type { MeshEndpointHook } from './hooks.js';
import { MAX_ADMISSION_PAYLOAD_SIZE, MAX_SERVICES_IN_ADMISSION, validateHexField } from '../limits.js';
import type { NonceStore } from './nonce.js';
import { createPeerAdmission, type PeerAttributeStore } from '../peer-store.js';

/**
 * Service summary returned in admission response.
 *
 * NOTE: Wire format uses snake_case keys (contract_id) for Python interop.
 * The camelCase interface fields here need mapping when parsing from wire.
 * TODO: Add proper wire-format mapping for ServiceSummary (pre-existing gap).
 */
export interface ServiceSummary {
  name: string;
  version: number;
  contractId: string;
  pattern: string;
  methods: string[];
  channels?: Record<string, string>;
}

/** Consumer admission request. */
export interface ConsumerAdmissionRequest {
  credentialJson: string;
  iidToken?: string;
}

/** Consumer admission response from producer. */
export interface ConsumerAdmissionResponse {
  admitted: boolean;
  reason?: string;
  services: ServiceSummary[];
  registryNamespace?: string;
  attributes?: Record<string, string>;
  rootPubkey?: string;
  /** Hex-encoded 32-byte gossip topic — only populated for root node. */
  gossipTopic?: string;
}

/** Options for server-side consumer admission handlers. */
export interface ConsumerAdmissionOpts {
  nonceStore?: NonceStore;
  services?: ServiceSummary[];
  registryNamespace?: string;
  allowUnenrolled?: boolean;
  /** Gossip topic ID (32 bytes). Included in response only for root node. */
  gossipTopicId?: Uint8Array;
  /** When supplied, successful admissions are recorded so the RPC dispatch
   *  layer can read the peer's attributes from CallContext. */
  peerStore?: PeerAttributeStore;
  logger?: { info(msg: string, ...args: unknown[]): void; warn(msg: string, ...args: unknown[]): void; error(msg: string, ...args: unknown[]): void };
}

// ── Client-side ───────────────────────────────────────────────────────────────

/**
 * Perform the consumer admission handshake.
 *
 * @param connection - The QUIC connection to the producer (admission ALPN)
 * @param credential - The consumer enrollment credential
 * @param iidToken - Optional cloud instance identity token
 * @returns The admission response with services and registry namespace
 */
export async function performAdmission(
  connection: { openBi(): Promise<{ takeSend(): any; takeRecv(): any }> },
  credential: ConsumerEnrollmentCredential,
  iidToken?: string,
): Promise<ConsumerAdmissionResponse> {
  const bi = await connection.openBi();
  const send = bi.takeSend();
  const recv = bi.takeRecv();

  // Build and send request (snake_case keys for Python interop).
  // The credential is serialised to its snake_case wire form via
  // consumerCredToJson — JSON.stringify on the camelCase TS interface
  // would produce the wrong keys and the server would reject it.
  const wireRequest = {
    credential_json: credential ? consumerCredToJson(credential) : '',
    iid_token: iidToken ?? '',
  };
  const reqBytes = new TextEncoder().encode(JSON.stringify(wireRequest));
  if (reqBytes.byteLength > MAX_ADMISSION_PAYLOAD_SIZE) {
    throw new Error(`admission request too large: ${reqBytes.byteLength} > ${MAX_ADMISSION_PAYLOAD_SIZE}`);
  }

  // Write request + finish send side
  await send.writeAll(reqBytes);
  await send.finish();

  // Read response (snake_case keys from Python wire format)
  const respBytes = await recv.readToEnd(MAX_ADMISSION_PAYLOAD_SIZE);
  const respText = new TextDecoder().decode(respBytes);
  if (!respText || respText.length === 0) {
    throw new Error('admission failed: server returned empty response (may be overloaded, retry)');
  }
  const d = JSON.parse(respText);
  const response: ConsumerAdmissionResponse = {
    admitted: d.admitted,
    reason: d.reason,
    services: d.services ?? [],
    registryNamespace: d.registry_namespace ?? d.registryNamespace ?? '',
    attributes: d.attributes ?? {},
    rootPubkey: d.root_pubkey ?? d.rootPubkey ?? '',
    gossipTopic: d.gossip_topic ?? d.gossipTopic,
  };

  // Validate
  if (response.services && response.services.length > MAX_SERVICES_IN_ADMISSION) {
    throw new Error(`admission response has ${response.services.length} services, max is ${MAX_SERVICES_IN_ADMISSION}`);
  }

  return response;
}

// ── Credential serialisation helpers ──────────────────────────────────────────

/**
 * Serialise a ConsumerEnrollmentCredential to the wire JSON format.
 * Hex-encodes rootPubkey, nonce, and signature fields.
 */
export function consumerCredToJson(cred: ConsumerEnrollmentCredential): string {
  return JSON.stringify({
    credential_type: cred.credentialType,
    root_pubkey: cred.rootPubkey,  // already hex in TS type
    expires_at: cred.expiresAt,
    attributes: cred.attributes,
    endpoint_id: cred.endpointId ?? null,
    nonce: cred.nonce ?? null,
    signature: cred.signature,
  });
}

/**
 * Deserialise a ConsumerEnrollmentCredential from the wire JSON format.
 * Validates hex field lengths (pubkey=64, nonce=64, signature=128 hex chars).
 */
export function consumerCredFromJson(json: string): ConsumerEnrollmentCredential {
  const d = JSON.parse(json);

  // Validate hex field lengths
  validateHexField('root_pubkey', d.root_pubkey ?? '');
  const nonceHex: string = d.nonce ?? '';
  if (nonceHex) {
    validateHexField('nonce', nonceHex);
  }
  const sigHex: string = d.signature ?? '';
  if (sigHex) {
    validateHexField('signature', sigHex);
  }
  const eid: string = d.endpoint_id ?? '';
  if (eid) {
    validateHexField('endpoint_id', eid);
  }

  return {
    credentialType: d.credential_type,
    rootPubkey: d.root_pubkey,
    expiresAt: Number(d.expires_at),
    attributes: d.attributes ?? {},
    endpointId: d.endpoint_id || undefined,
    nonce: nonceHex || undefined,
    signature: sigHex || '',
  };
}

// ── Server-side handler ─────────────────────────────────────────────────────

/**
 * Server-side handler for the aster.consumer_admission ALPN.
 *
 * @param requestJson - JSON-serialised ConsumerAdmissionRequest.
 * @param rootPubkey - The server's root public key (hex string, 64 chars).
 * @param hook - MeshEndpointHook; addPeer is called on successful admission.
 * @param peerNodeId - QUIC peer identity from the connection handshake.
 * @param opts - Additional options (nonceStore, services, registryNamespace, allowUnenrolled, logger).
 * @returns ConsumerAdmissionResponse — always returned, never throws.
 */
export async function handleConsumerAdmissionRpc(
  requestJson: string,
  rootPubkey: string,
  hook: MeshEndpointHook,
  peerNodeId: string,
  opts: ConsumerAdmissionOpts = {},
): Promise<ConsumerAdmissionResponse> {
  const log = opts.logger ?? console;
  const denied: ConsumerAdmissionResponse = {
    admitted: false,
    reason: '', // oracle protection — never leak reason on wire
    services: [],
    rootPubkey,
  };

  // Parse the outer request envelope
  let req: ConsumerAdmissionRequest;
  try {
    const parsed = JSON.parse(requestJson);
    req = {
      credentialJson: parsed.credentialJson ?? parsed.credential_json ?? '',
      iidToken: parsed.iidToken ?? parsed.iid_token ?? '',
    };
  } catch (err) {
    log.warn(`consumer admission: malformed request from ${peerNodeId}: ${err}`);
    return denied;
  }

  // Include gossip topic only when the connecting peer IS the root node
  // (its endpoint_id == root_pubkey hex). This lets the operator's shell
  // observe the producer mesh without exposing the topic to other consumers.
  let topicForPeer = '';
  if (opts.gossipTopicId && peerNodeId === rootPubkey) {
    topicForPeer = bytesToHex(opts.gossipTopicId);
    log.info('consumer admission: root node detected — including gossip topic');
  }

  // Dev mode / open gate: empty credential -> auto-admit
  const isEmptyCredential = !req.credentialJson || req.credentialJson === '{}' || req.credentialJson === 'null';
  if (isEmptyCredential && opts.allowUnenrolled) {
    hook.addPeer(peerNodeId);
    if (opts.peerStore) {
      opts.peerStore.admit(createPeerAdmission({
        endpointId: peerNodeId,
        attributes: new Map(),
        admissionPath: 'aster.consumer_admission',
      }));
    }
    const role = topicForPeer ? 'root' : 'open gate';
    log.info(`consumer admission: auto-admitted ${peerNodeId} (${role})`);
    return {
      admitted: true,
      attributes: {},
      services: opts.services ?? [],
      registryNamespace: opts.registryNamespace ?? '',
      rootPubkey,
      gossipTopic: topicForPeer || undefined,
      reason: '',
    };
  }

  // Parse the inner credential
  let cred: ConsumerEnrollmentCredential;
  try {
    cred = consumerCredFromJson(req.credentialJson);
  } catch (err) {
    log.warn(`consumer admission: malformed credential from ${peerNodeId}: ${err}`);
    return denied;
  }

  // Trust anchor check: credential's rootPubkey must match server's
  if (!cred.rootPubkey || cred.rootPubkey !== rootPubkey) {
    log.warn(
      `consumer admission: untrusted root key from ${peerNodeId} ` +
      `(got ${(cred.rootPubkey ?? '(none)').slice(0, 12)}, expected ${rootPubkey.slice(0, 12)})`,
    );
    return denied;
  }

  // Run admission checks (offline + runtime)
  const result = await admit(cred, peerNodeId, {
    nonceStore: opts.nonceStore,
    iidToken: req.iidToken || undefined,
  });

  if (!result.admitted) {
    log.info(`consumer admission: denied ${peerNodeId}`);
    return denied;
  }

  hook.addPeer(peerNodeId);
  if (opts.peerStore) {
    const attrMap = new Map<string, string>();
    for (const [k, v] of Object.entries(result.attributes ?? {})) {
      attrMap.set(k, String(v));
    }
    opts.peerStore.admit(createPeerAdmission({
      endpointId: peerNodeId,
      attributes: attrMap,
      admissionPath: 'aster.consumer_admission',
    }));
  }
  log.info(`consumer admission: admitted ${peerNodeId}`);

  return {
    admitted: true,
    attributes: result.attributes ?? {},
    services: opts.services ?? [],
    registryNamespace: opts.registryNamespace ?? '',
    rootPubkey,
    gossipTopic: topicForPeer || undefined,
    reason: '',
  };
}

// ── Per-connection handler ──────────────────────────────────────────────────

/**
 * Handle one consumer admission connection: read request, write response.
 *
 * @param conn - A QUIC connection with acceptBi() and remoteId() methods.
 * @param rootPubkey - Hex-encoded root public key.
 * @param hook - MeshEndpointHook for peer admission tracking.
 * @param opts - Additional options.
 */
export async function handleConsumerAdmissionConnection(
  conn: {
    acceptBi(): Promise<{ takeSend(): any; takeRecv(): any }>;
    remoteId(): string;
  },
  rootPubkey: string,
  hook: MeshEndpointHook,
  opts: ConsumerAdmissionOpts = {},
): Promise<void> {
  const peerNodeId = conn.remoteId();
  const log = opts.logger ?? console;
  try {
    const bi = await conn.acceptBi();
    const send = bi.takeSend();
    const recv = bi.takeRecv();

    const raw: Uint8Array = await recv.readToEnd(MAX_ADMISSION_PAYLOAD_SIZE);
    if (!raw || raw.length === 0) {
      log.warn(`consumer admission: empty request from ${peerNodeId}`);
      return;
    }

    const requestJson = new TextDecoder().decode(raw);

    const response = await handleConsumerAdmissionRpc(
      requestJson,
      rootPubkey,
      hook,
      peerNodeId,
      opts,
    );

    // Serialise response — snake_case keys for Python interop.
    // Strip reason on wire (oracle protection).
    // Convert ServiceSummary to snake_case for wire compat
    const wireServices = (response.services ?? []).map((s: any) => ({
      name: s.name,
      version: s.version,
      contract_id: s.contractId ?? s.contract_id ?? '',
      pattern: s.pattern,
      methods: s.methods,
      channels: s.channels ?? {},
    }));
    const wireResponse: Record<string, unknown> = {
      admitted: response.admitted,
      attributes: response.attributes ?? {},
      services: wireServices,
      registry_namespace: response.registryNamespace ?? '',
      root_pubkey: response.rootPubkey ?? '',
      reason: '', // never leak reason on wire
    };
    if (response.gossipTopic) {
      wireResponse.gossip_topic = response.gossipTopic;
    }

    const responseBytes = new TextEncoder().encode(JSON.stringify(wireResponse));
    try {
      await send.writeAll(responseBytes);
      await send.finish();
    } catch (writeErr) {
      log.warn(`consumer admission: failed to send response to ${peerNodeId}: ${writeErr}`);
      // Stream write failed — client will see an empty/closed stream.
      // Nothing more we can do; the connection may already be gone.
    }
    // Don't conn.close() — let QUIC drain the streams naturally.
    // Calling close() sends CONNECTION_CLOSE which kills in-flight
    // data before the consumer can readToEnd().
  } catch (err) {
    log.warn(`consumer admission: error handling ${peerNodeId}: ${err}`);
  }
}

// ── Accept loop ─────────────────────────────────────────────────────────────

/**
 * Accept and process connections on aster.consumer_admission until cancelled.
 *
 * Runs as a background task alongside the main server. Each connection is
 * handled concurrently so one slow consumer cannot block others.
 *
 * @param node - An endpoint/node bound to the consumer admission ALPN, with accept().
 * @param rootPubkey - Hex-encoded root public key.
 * @param hook - MeshEndpointHook allowlist manager.
 * @param opts - Additional options.
 */
export async function serveConsumerAdmission(
  node: { accept(): Promise<any> },
  rootPubkey: string,
  hook: MeshEndpointHook,
  opts: ConsumerAdmissionOpts = {},
): Promise<void> {
  const log = opts.logger ?? console;
  try {
    while (true) {
      const conn = await node.accept();
      // Fire-and-forget: handle each connection concurrently
      handleConsumerAdmissionConnection(conn, rootPubkey, hook, opts).catch((err) => {
        log.warn(`consumer admission: connection handler error: ${err}`);
      });
    }
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    if (msg.includes('abort') || msg.includes('cancel')) return;
    log.error(`serveConsumerAdmission: unexpected error: ${msg}`);
  }
}

File: bindings/python/aster/trust/consumer.py
"""
aster.trust.consumer — Consumer admission handler and wire types.

Spec reference: Aster-trust-spec.md §3.2, §3.2.2.

Two actors:

  Server (producer node)
    ``handle_consumer_admission_rpc`` — verifies credential, admits peer,
    returns ``ConsumerAdmissionResponse`` with services + registry_ticket.

  Client (consumer node)
    ``ConsumerAdmissionRequest`` — the wire message to send.
    ``ConsumerAdmissionResponse.from_json`` — parse the server's reply.

Wire format: newline-delimited JSON over a QUIC bidi-stream on
``aster.consumer_admission`` ALPN.  The client sends one JSON line; the
server responds with one JSON line and closes the stream.
"""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import dataclass, field
from typing import Callable, TYPE_CHECKING

from .admission import admit
from .credentials import ConsumerEnrollmentCredential

if TYPE_CHECKING:
    from .hooks import MeshEndpointHook
    from .nonces import NonceStoreProtocol
    from ..registry.models import ServiceSummary

logger = logging.getLogger(__name__)


# ── Wire types ────────────────────────────────────────────────────────────────


@dataclass
class ConsumerAdmissionRequest:
    """Wire message sent by a consumer over ``aster.consumer_admission``.

    ``credential_json`` is a JSON object with these keys:
      credential_type, root_pubkey (hex), expires_at (int), attributes (dict),
      endpoint_id (str | null), nonce (hex | null), signature (hex).
    ``iid_token`` is optional; empty string if not provided.
    """

    credential_json: str
    iid_token: str = ""

    def to_json(self) -> str:
        return json.dumps(
            {"credential_json": self.credential_json, "iid_token": self.iid_token},
            separators=(",", ":"),
        )

    @classmethod
    def from_json(cls, raw: str | bytes) -> "ConsumerAdmissionRequest":
        d = json.loads(raw)
        return cls(
            credential_json=d["credential_json"],
            iid_token=d.get("iid_token") or "",
        )


@dataclass
class ConsumerAdmissionResponse:
    """Response returned by the server after consumer admission (§3.2.2).

    ``admitted=True`` on success.  ``reason`` is always empty in the wire
    response (oracle protection); it is only set for local diagnostics.
    """

    admitted: bool
    attributes: dict[str, str] = field(default_factory=dict)
    services: list["ServiceSummary"] = field(default_factory=list)
    registry_ticket: str = ""
    root_pubkey: str = ""        # hex-encoded 32-byte ed25519 public key
    gossip_topic: str = ""       # hex-encoded 32-byte topic — only for root node
    reason: str = ""             # MUST be empty in wire response (§3.2.2)

    def to_json(self) -> str:
        from ..registry.models import ServiceSummary as _SS  # avoid circular at module level

        d = {
            "admitted": self.admitted,
            "attributes": self.attributes,
            "services": [
                s.to_json_dict() if isinstance(s, _SS) else s
                for s in self.services
            ],
            "registry_ticket": self.registry_ticket,
            "root_pubkey": self.root_pubkey,
            "reason": "",   # never leak reason on wire
        }
        if self.gossip_topic:
            d["gossip_topic"] = self.gossip_topic
        return json.dumps(d, separators=(",", ":"))

    @classmethod
    def from_json(cls, raw: str | bytes) -> "ConsumerAdmissionResponse":
        from ..registry.models import ServiceSummary as _SS

        from ..limits import MAX_SERVICES_IN_ADMISSION, MAX_CHANNELS_PER_SERVICE

        d = json.loads(raw)
        raw_services = d.get("services") or []
        if len(raw_services) > MAX_SERVICES_IN_ADMISSION:
            raw_services = raw_services[:MAX_SERVICES_IN_ADMISSION]
        services = [
            _SS.from_json_dict(s) if isinstance(s, dict) else s
            for s in raw_services
        ]
        return cls(
            admitted=bool(d["admitted"]),
            attributes=d.get("attributes") or {},
            services=services,
            registry_ticket=d.get("registry_ticket") or "",
            root_pubkey=d.get("root_pubkey") or "",
            gossip_topic=d.get("gossip_topic") or "",
            reason="",
        )


# ── Credential serialisation helpers ─────────────────────────────────────────


def consumer_cred_to_json(cred: ConsumerEnrollmentCredential) -> str:
    """Serialise a ConsumerEnrollmentCredential to the wire JSON format."""
    return json.dumps(
        {
            "credential_type": cred.credential_type,
            "root_pubkey": cred.root_pubkey.hex(),
            "expires_at": cred.expires_at,
            "attributes": cred.attributes,
            "endpoint_id": cred.endpoint_id,
            "nonce": cred.nonce.hex() if cred.nonce is not None else None,
            "signature": cred.signature.hex(),
        },
        separators=(",", ":"),
    )


def consumer_cred_from_json(s: str | bytes) -> ConsumerEnrollmentCredential:
    """Deserialise a ConsumerEnrollmentCredential from the wire JSON format."""
    from ..limits import validate_hex_field

    d = json.loads(s)

    # Validate hex field lengths before parsing
    validate_hex_field("root_pubkey", d.get("root_pubkey", ""))
    nonce_hex = d.get("nonce") or ""
    if nonce_hex:
        validate_hex_field("nonce", nonce_hex)
    sig_hex = d.get("signature") or ""
    if sig_hex:
        validate_hex_field("signature", sig_hex)
    eid = d.get("endpoint_id") or ""
    if eid:
        validate_hex_field("endpoint_id", eid)

    return ConsumerEnrollmentCredential(
        credential_type=d["credential_type"],
        root_pubkey=bytes.fromhex(d["root_pubkey"]),
        expires_at=int(d["expires_at"]),
        attributes=d.get("attributes") or {},
        endpoint_id=d.get("endpoint_id"),
        nonce=bytes.fromhex(nonce_hex) if nonce_hex else None,
        signature=bytes.fromhex(sig_hex) if sig_hex else b"",
    )


# ── Server-side handler ───────────────────────────────────────────────────────


async def handle_consumer_admission_rpc(
    request_json: str,
    root_pubkey: bytes,
    hook: "MeshEndpointHook",
    peer_node_id: str,
    nonce_store: "NonceStoreProtocol | None" = None,
    services: "list[ServiceSummary] | None" = None,
    registry_ticket: str = "",
    allow_unenrolled: bool = False,
    gossip_topic_id: bytes | None = None,
) -> ConsumerAdmissionResponse:
    """Server-side handler for the ``aster.consumer_admission`` ALPN.

    Args:
        request_json:     JSON-serialised ``ConsumerAdmissionRequest``.
        root_pubkey:      The server's root public key (32 bytes) used to
                          verify the credential's signature.
        hook:             ``MeshEndpointHook`` — ``add_peer`` is called on
                          successful admission.
        peer_node_id:     QUIC peer identity from the connection handshake.
        nonce_store:      Required when accepting OTT credentials.
        services:         List of ``ServiceSummary`` to include in the response.
                          Pass ``None`` or ``[]`` when no services are published.
        registry_ticket:  Read-only iroh-docs share ticket for the registry doc;
                          empty string if this node does not operate a registry.
        allow_unenrolled: When True, auto-admit peers that present an empty
                          credential (dev mode / ``allow_all_consumers=True``).

    Returns:
        ``ConsumerAdmissionResponse`` — always returned, never raises.
        On failure ``admitted=False`` and ``reason=""`` (no oracle leak).
    """
    _denied = ConsumerAdmissionResponse(
        admitted=False,
        root_pubkey=root_pubkey.hex(),
    )

    try:
        req = ConsumerAdmissionRequest.from_json(request_json)
    except Exception as exc:  # noqa: BLE001
        logger.warning("consumer admission: malformed request from %s: %s", peer_node_id, exc)
        return _denied

    from aster.health import get_admission_metrics
    _adm = get_admission_metrics()

    # Include gossip topic only when the connecting peer IS the root node
    # (its endpoint_id == root_pubkey hex). This lets the operator's shell
    # observe the producer mesh without exposing the topic to other consumers.
    _topic_for_peer = ""
    if gossip_topic_id and peer_node_id == root_pubkey.hex():
        _topic_for_peer = gossip_topic_id.hex()
        logger.debug("consumer admission: root node detected — including gossip topic")

    # Dev mode / open gate: empty credential → auto-admit.
    if not req.credential_json and allow_unenrolled:
        if hook is not None:
            hook.add_peer(peer_node_id)
        _adm.record_consumer_admit()
        _role = "root" if _topic_for_peer else "open gate"
        logger.info("consumer admission: auto-admitted %s (%s)", peer_node_id, _role)
        return ConsumerAdmissionResponse(
            admitted=True,
            attributes={},
            services=list(services or []),
            registry_ticket=registry_ticket,
            root_pubkey=root_pubkey.hex(),
            gossip_topic=_topic_for_peer,
        )

    try:
        cred = consumer_cred_from_json(req.credential_json)
    except Exception as exc:  # noqa: BLE001
        logger.warning("consumer admission: malformed credential from %s: %s", peer_node_id, exc)
        return _denied

    # Verify the credential's root_pubkey matches the server's trusted key.
    # Without this check any party could mint their own root key and self-admit.
    if cred.root_pubkey != root_pubkey:
        logger.warning(
            "consumer admission: untrusted root key from %s (got %s, expected %s)",
            peer_node_id,
            cred.root_pubkey.hex()[:12],
            root_pubkey.hex()[:12],
        )
        return _denied

    result = await admit(
        cred,
        peer_node_id,
        nonce_store=nonce_store,
        iid_token=req.iid_token or None,
    )

    if not result.admitted:
        _adm.record_consumer_deny()
        logger.info("consumer admission: denied %s", peer_node_id)
        return _denied

    if hook is not None:
        hook.add_peer(peer_node_id)
    _adm.record_consumer_admit()
    logger.info("consumer admission: admitted %s", peer_node_id)

    return ConsumerAdmissionResponse(
        admitted=True,
        attributes=result.attributes or {},
        services=list(services or []),
        registry_ticket=registry_ticket,
        root_pubkey=root_pubkey.hex(),
        gossip_topic=_topic_for_peer,
    )


# ── Listener helper ───────────────────────────────────────────────────────────


async def serve_consumer_admission(
    endpoint: object,
    root_pubkey: bytes,
    hook: "MeshEndpointHook",
    nonce_store: "NonceStoreProtocol | None" = None,
    services_getter: "Callable[[], list[ServiceSummary]] | None" = None,
    registry_ticket_getter: "Callable[[], str] | None" = None,
) -> None:
    """Accept and process connections on ``aster.consumer_admission`` until cancelled.

    Runs as a background task alongside the main server.  Each connection is
    handled in its own ``asyncio.Task`` so one slow consumer cannot block others.

    Args:
        endpoint:               A ``NetClient`` (from ``create_endpoint_with_config``)
                                bound to the ``aster.consumer_admission`` ALPN.
        root_pubkey:            The server's root public key (32 bytes).
        hook:                   ``MeshEndpointHook`` allowlist manager.
        nonce_store:            Required for OTT credentials.
        services_getter:        Callable returning current ``list[ServiceSummary]``.
                                Called fresh for every admission.
        registry_ticket_getter: Callable returning the current doc share ticket.
                                Called fresh for every admission.
    """
    try:
        while True:
            conn = await endpoint.accept()
            asyncio.create_task(
                handle_consumer_admission_connection(
                    conn,
                    root_pubkey=root_pubkey,
                    hook=hook,
                    nonce_store=nonce_store,
                    services_getter=services_getter,
                    registry_ticket_getter=registry_ticket_getter,
                )
            )
    except asyncio.CancelledError:
        pass
    except Exception as exc:  # noqa: BLE001
        logger.error("serve_consumer_admission: unexpected error: %s", exc)


async def handle_consumer_admission_connection(
    conn: object,
    root_pubkey: bytes,
    hook: "MeshEndpointHook",
    nonce_store: "NonceStoreProtocol | None" = None,
    services_getter: "Callable[[], list[ServiceSummary]] | None" = None,
    registry_ticket_getter: "Callable[[], str] | None" = None,
    allow_unenrolled: bool = False,
    gossip_topic_getter: "Callable[[], bytes | None] | None" = None,
) -> None:
    """Handle one consumer admission connection: read request, write response."""
    peer_node_id = conn.remote_id()
    try:
        send, recv = await conn.accept_bi()
        raw = await recv.read_to_end(64 * 1024)
        if not raw:
            logger.warning("consumer admission: empty request from %s", peer_node_id)
            return

        services = services_getter() if services_getter is not None else []
        ticket = registry_ticket_getter() if registry_ticket_getter is not None else ""
        topic = gossip_topic_getter() if gossip_topic_getter is not None else None

        response = await handle_consumer_admission_rpc(
            raw.decode(),
            root_pubkey=root_pubkey,
            allow_unenrolled=allow_unenrolled,
            hook=hook,
            peer_node_id=peer_node_id,
            nonce_store=nonce_store,
            services=services,
            registry_ticket=ticket,
            gossip_topic_id=topic,
        )

        await send.write_all(response.to_json().encode())
        await send.finish()
        # Don't conn.close() — let QUIC drain the streams naturally.
        # Calling close() sends CONNECTION_CLOSE which kills in-flight
        # data before the consumer can read_to_end().
    except Exception as exc:  # noqa: BLE001
        logger.warning("consumer admission: error handling %s: %s", peer_node_id, exc)
